### DISCUSSION TITLE
Supporting IBC Casting

### DISCUSSION THREAD

**Discussion by michaelfig** [cosmos/ibc#759] [Ideas]
# Motivation

The goal of IBC Casting is to support the model of:
- a single publishing IBC host
- many following clients, with no required submission of transactions to a chain
- a single Merkle proof for a given "stream head" is sufficient to link together the entire history of published packets

## Implementations

- Agoric's Golang Casting streamer at https://github.com/Agoric/agoric-sdk/blob/mfig-vstream/golang/cosmos/x/vstream/spec/01_concepts.md#concepts (PR at Agoric/agoric-sdk#5466)
- Agoric's (off-chain) Javascript Casting follower at https://github.com/Agoric/agoric-sdk/tree/master/packages/casting/README.md

## IBC PubSub Discussion History

**This is the unchanged initial post made by @michaelfig introducing IBC PubSub**

I had an idea for a new IBC-based feature grounded in actual use cases on the Agoric chain.

Could we make a new connection primitive (other than the bidirectional "channel") for publish/subscribe applications?  The API I envision is for a source host* to be able to publish messages on a given topic, and another host/client** to subscribe to that topic. 
Unlike IBC channels, a pubsub would be unidirectional and not have any means to send data from the subscriber to the publisher (not even to handshake).  Here are some loose requirements I'm thinking of:

`*` IBC host = chain or solo.
`**` including relayerless client libraries like CosmJS.

1. Topics need to have well-known names, and the namespace may be much larger than the current port IDs.
2. Publishers open a descriptor for a single topic per host, and can publish messages at any time.
3. Only the messages for the most recent published block needs to be in the current chain state.  This is to eliminate the need for acks to clear old messages. If message history is desired, it can be obtained via historical queries.
4. Subscriptions name a connection and topic and may be multihop.  This is not a shared bus, it is messages pulled by relayers from a single publisher to a specific subscriber.
5. An individual subscriber can open a subscription requester for a given publisher's IBC connection, and use it to create requests for relayers to fulfill with the last published message on given topic after given publisher block height.  Maybe return all the messages in that block?  Maybe separately?
6. Relayer incentivisation would be done as part of the subscription request, once the relayer can demonstrate that the relayed message satisfies the request.
 
I believe the provable data needed to store on the publishing side is per-host, per-topic, and consists of information about the last block that published something:
- block height
- array of messages (each containing base64 data) published in that block
- block height of prior publishing block (to be able to walk backwards through history)

Thoughts?

Tagging: @dtribble, @erights, @jackzampolin, @JimLarson, @warner, @zmanian for moral support.

**Comment by michaelfig** [comment]
Two more folks: @AdityaSripal, @dckc.

**Comment by zmanian** [comment]
Seems related to IBC queries.
https://github.com/cosmos/ibc/pull/735

**Comment by michaelfig** [comment]
I agree that it is related, to IBC queries.  Let me explain the differences:

1. Pubsub has a predicate to allow subscription to a future value (using events as gossip to announce an update) but queries are polling.
2. Pubsub handles streamed packets and preserved history where queries are based on single values.

I could implement pubsub as an application on top of queries, but only if not polling, and if the pubsub application could provide predicates to filter the acceptable query answers.

For this reason, I see pubsub as closer to a connection protocol than to a polling one-shot query.

@zmanian do you personally think it makes sense to add supporting features to ibc query?

**Reply by zmanian** [reply]
I think queries just need to be able to support a stream of values emitted over time rather than just a single value and then we can fuse both .

The infrastructure of pubsub and queries should be identical

**Comment by michaelfig** [comment]
I've figured out two orthogonal but composable features for Interchain Queries that would enable IBC PubSub to be built as an application:
    - add a "different-than V" query limiter that requires the result not to match `V` in order to satisfy the query
    - add an "after-height H" query limiter that can only satisfy the query with a result in some block *after* blockheight `H`
    - compose them to get "different-than `V` AND after-height `H`"

Implementation is opt-in for relayers.  Worst case, a relayer would poll periodically to see if new data arrives.  If the source chain can raise on-change events for query results, that can help a relayer or other off-chain follower to satisfy the query more promptly/efficiently.

Here's a suggestion for change events:
- `ibcpub.path=<store>:<key>` the store and key's data has changed
- `ibcpub.data=<base64-encoding, or 'null'>` the new contents of the path

The `ibcpub.data` event is optional, and can be used as gossip for trusting clients (like simple event processors) to obtain the data quickly without having to issue a separate proved query.

**Comment by michaelfig** [comment]
I'm working on a PR at Agoric/agoric-sdk#5466 !  You're welcome to join.

**Comment by ethanfrey** [comment]
Interesting idea. What I would like to ask is the delivery semantics, which is one of the trickiest details of any messaging bus.

IBC channels guarantee "at most once" delivery. But also work hard to ensure that a response or timeout will eventually arrive on the sender, if the channel/connection remains alive (and push any liveness requirements to that lower level).

What about pubsub? If I am a subscriber, can I assume that I will get all messages published after my subscription is registered on the remote chain, until their is a liveness issue on the connection? If there are messages [A, B, C, D] published, and a subscriber connects and receives B as the first message - will it ever see A? Is it possible for it to receive only [B, D] and not C? Are there any ordering guarantees?

These are very important semantics to define before implementing anything. I am not sure what the right choices are to support the most applications, but for me allowing `[B, D]` to ever be valid seems like it opens up a lot of potential to abuse by a malicious relayer

**Reply by michaelfig** [reply]
The back pointers ("prior position" in my latest PR updates) allow any client who receives `[B, D]` to also reconstruct `[A, B, C, D]`.

I think the [lossless forward iteration algorithm](https://github.com/Agoric/agoric-sdk/blob/mfig-vstream/golang/cosmos/x/vstream/spec/01_concepts.md#forward-iteration-lossless-history) may be what you're looking for.

**Reply by ethanfrey** [reply]
I'm not talking about algorithms to implement.
I mean what is the desired guarantees to subscribers? Assuming mischievous relayers

**Reply by michaelfig** [reply]
Ah, I see.  I'm talking about guarantees in the face of the particular supported algorithms.  The reverse queries and point-in-time queries are fairly easy to reason about.  That's why I focused on lossless forward iteration.

For lossless forward iteration, A mischievous relayer can always refuse to return anything for a loosely-constrained (i.e. "store-key K, different-than V, later-than-height H") interchain query.  But if a relayer returns any constrained value, it satisfies the only constraint needed by a subscriber.

After any future value, `D`, the subscriber will see that the sequence number is not continuous, and will have the information via the back pointer to begin issuing more, tightly-constrained queries (i.e. "store-key K, exact-height H") to backfill the missing history.  Relayers can deny that service, but for that backfilling phase they cannot return just any height they want.

And all of these query results must be provable during the latest unbonding period.

I'd love to engage in better threat modelling than this, but does it give you at least a hint of what you're asking for wrt guarantees?

**Reply by ethanfrey** [reply]
This covers the main points. It would be good to write this more formally and show the guarantees and expected behaviours of a client. 

Mainly that they cannot treat absence of evidence as evidence as absence (there is no proof of absence, right?). In particular, this means should not make any assumptions on the state if they have not received an update within some time X.

And also that they cannot expect all values to be continuous, but can detect if there is a gap in the messages.

**Reply by ethanfrey** [reply]
Honestly, this seems very poor guarantees to push to the app level, and if this should be a core protocol (akin to an unordered channel), it would be nice to provide a higher level primitive to apps.

In particular, a proof that no change between height H and X, where H is the requested height and X is a more recent block header from the remote chain (exposing the timestamp as well as height).

From that, you could build some message to the app that says "no feed in 12 hours" for example, to distinguish between no state changes and no relayers running. 

> I had an idea for a new IBC-based feature grounded in actual use cases on the Agoric chain.

Maybe it would help to see the use cases you want to build with this (I didn't see any description in those links). One case I thought of was mirroring the state of an oracle or a multisig from one chain to others. A multisig may remain unchanged for days. But you do want to freeze the copy if you cannot prove you have a reasonably recent state.

**Reply by michaelfig** [reply]
> In particular, a proof that no change between height H and X, where H is the requested height and X is a more recent block header from the remote chain (exposing the timestamp as well as height).
> 
> From that, you could build some message to the app that says "no feed in 12 hours" for example, to distinguish between no state changes and no relayers running.

That's a terrific idea!  Yes, that way a long delay between changes won't fall outside of the unbonding period.  The relayer could return proof that the value has not changed at the end of the deadline, and then the client can reissue a fresh query for another span even further in the future ("different-than V, later-than-block X, timeout-at-block Y").

(Documentation updated in the [interchain query section](https://github.com/Agoric/agoric-sdk/blob/mfig-vstream/golang/cosmos/x/vstream/spec/01_concepts.md#forward-iteration-lossless-history))

> Maybe it would help to see the use cases you want to build with this (I didn't see any description in those links).

(Please forgive my sketchiness, I'm in the process of reconstructing and documenting an under-documented mechanism we currently have.)

I added a link to the `@agoric/chain-streams` library [MVP application description](https://github.com/Agoric/agoric-sdk/blob/mfig-chain-streams/packages/chain-streams/README.md).    Basically, we currently have a hacky way to publish the full contents of a "client mailbox" at a given Cosmos store key (currently abstracted as a dot-separated "StoragePath"), and want to iterate with a given integrity level without sending any transactions to the chain.

It would be great to be able to upgrade that mechanism to use a more standard lossless forward iteration (taking advantage of archive nodes).  For example, we could have another smart contract chain (such as another Agoric JS-based or CosmWasm-based one) contract participating as an Agoric client via Interchain queries and Interchain Accounts.

> One case I thought of was mirroring the state of an oracle or a multisig from one chain to others. A multisig may remain unchanged for days. But you do want to freeze the copy if you cannot prove you have a reasonably recent state.

I don't understand this application fully enough.

**Reply by AdityaSripal** [reply]
Agreed with @ethanfrey that we should retain the guarantee from IBC that relayers are only reliant on liveness and do not have a free hand in choosing values.

Realistically people will only have one model in their head about relayers and move forward on that assumption. I'm wary of relaxing that property in the query case, because people will certainly not realize this and they can press forward with use cases that might be gamable by relayers without realizing it. (ie. lotteries, price oracles, etc).

I would prefer that we have the following scheme:

If a chain wishes to provide PubSub for a particular key, they **must** store the value it in the following format:

```
{
    value: current_value
    current_height: height
    prev_value: previous_vale
    prev_height: prev_height
}
```

With each update to the value, the current value and height get moved to previous and the new values are stored in current

Now suppose the querier already knows that value is `V` at block `X`. It just wants to be notified of the next update.

Issue query: `different than V, later than block X, timeout at block Y`

1. If value changes between X < Y

Relayer can submit a proof of this and the querier chain checks that the previous value is V and previous height is X. They then know that they are receiving the next update. Querying chain may do logic on this basis.

2. If value does not change until after Y.

Relayer can submit a proof at height > Y that the current value and height are still `V` and `X`. Querying chain may do logic on this basis.

3. If timeout occurs without a relayer message

Querying chain must not assume anything about the counterparty state machine. It may reissue query with new timeout.

With this I believe we support your usecase @michaelfig , but we do not allow untrusted relayers a free hand in choosing a particular value to the querying chain.

Please let me know if this works for you.

**Reply by michaelfig** [reply]
That works.

I have a more sophisticated design in progress, but it only relies on `different than V, later than block X, timeout at block Y` is supported. Just like there can be alternative token transfer protocols, my design can be an application of the underlying mechanism.

What I am concerned about at this time is getting that underlying mechanism in place so that there can be innovation higher up on the (PubSub) stack.

**Reply by michaelfig** [reply]
I guess I don't understand the "must" in "they *must* store the value it in the following format".  That "must" should only be as strong as "we will standardize this protocol and conforming implementations *must* store it as follows".  If I use a nonstandard or extended protocol, I'm free to choose my own layout.

**Reply by michaelfig** [reply]
Here is an updated semantics that would support my use case as well as yours:

Provide `prefix matches bytes P, later than block X, timeout at block Y`, then rearrange your data structure so that you have:

```js
{
    prev_value: previous_value,
    prev_height: prev_height,
    value: current_value,
    current_height: height,
}
```

and you would issue a query that used prefix matches bytes corresponding to `{ prev_value: previous_value, prev_height: prev_height,`.

How does that sound?

**Reply by ethanfrey** [reply]
Thank you @AdityaSripal for understanding my concerns and explaining it better than I could:

> We should retain the guarantee from IBC that relayers are only reliant on liveness and do not have a free hand in choosing values.
>
> Realistically people will only have one model in their head about relayers and move forward on that assumption. I'm wary of relaxing that property in the query case

Michael:

> What I am concerned about at this time is getting that underlying mechanism in place so that there can be innovation higher up on the (PubSub) stack.


The big question for me is "higher up in the stack".

We expose "order channels" and "unordered channels" to apps now. Anything else we expose should have some similar high-level guarantees.

Maybe these are two conversations. (1) what we expose to the app level (2) what do we implement in the "channel" level. Maybe we can reuse 90% of the code for "ordered pubsub" "out of order pubsub" "pubsub with guarantee X". But that should be built into the stack (TAO) and not pushed into app developers.

I personally think defining a few use cases to support is a great way to start, then dig into one mechanism that allows us to implement them with maximum code/logic reuse.

**Reply by michaelfig** [reply]
> I personally think defining a few use cases to support is a great way to start, then dig into one mechanism that allows us to implement them with maximum code/logic reuse.

Agreed.  The algorithms I listed in https://github.com/Agoric/agoric-sdk/blob/mfig-vstream/golang/cosmos/x/vstream/spec/01_concepts.md#consuming exactly define the use cases I'm trying to support.

I'm curious why nobody has commented on them specifically yet.

**Reply by ethanfrey** [reply]
Nice link, I tried another link at the top (there were 3) and didn't find it nearly as illuminating.

Those still describe protocols and how to implement those. I wonder if we need to expose this all to the app level. Especially dealing with liveness.

Also, while I can see uses for the first 3, I cannot see a clear use for reverse iteration over every state change. It would be nice to explain sample apps there.

**Reply by AdityaSripal** [reply]
@michaelfig Seems to me like the proposal I had above would cover the use cases in your document, modulo minor changes?

I think the key property I'm proposing to enforce is that relayers are not allowed to skip values in the stream.

**Reply by michaelfig** [reply]
You are correct: your proposal is good for understanding the concepts relevant to IBC clients and relayers.  However, I would encourage you to take a look at [the latest `stream.proto` file for my prototype implementation](.  It is my attempt to bring the general casting feature to production, not just for IBC relayers.

It accommodates a few key features that your proposal currently lacks:
- termination (including error conditions)
- forking of a stream from a position in an earlier stream
- multiple published values per block

I understand that the IBC-specific design is important to you.  The publisher/subscriber feature set is critical to me (and indeed, the non-IBC case is the initial motivation for the Agoric feature as a whole), and I added IBC support as an afterthought.

I'd like to see a proposed ICS standard with the relayer simplicity of your proposal merged with features of mine.

**Reply by michaelfig** [reply]
I forgot to attach the link to current prototype `stream.proto`: https://github.com/Agoric/agoric-sdk/blob/mfig-vstream/golang/cosmos/proto/agoric/types/stream.proto

**Reply by AdityaSripal** [reply]
Thank you! Attaching the streams.proto is indeed very helpful.

I think ideally we'd want to build the simplest version of queries such that your streams protocol can be built on top of it, even if we don't implement streams directly into the IBC part itself.

So given that IBC queries is data agnostic, the only thing that seems necessary to me to build into the IBC queries protocol is the the following feature: get next different value

The features you mention that is needed for streams:

> termination (including error conditions)
> multiple published values per block
> forking of a stream from a position in an earlier stream

These all seem like they can be built at the application level above queries. If IBC queries is just handling raw bytes stored in a particular key. Then it is up to the query caller and storer to decide on a mutually intelligible format.

So the mutually intelligible format could well be a `StreamCell`, that stores a pointer to the previous position, multiple values, and the rest. The streams application can then just use IBC queries, to get the next change in the key-value, decode to a StreamCell and interpret the change as being multiple updates or a termination or whatever other stream-specific update.

If you agree that we can build Streams and everything you need at the application layer sitting above IBC queries (so long as we have get next value), then I would be in favor of getting that into the current queries spec.

My goal is to make sure that we can support the widest range of use cases with the simplest approach. Let me know if this works for you, and we can get started on it! cc: @angbrav

**Comment by AdityaSripal** [comment]
Thanks for the idea @michaelfig! Agreed with @zmanian that we should try to merge efforts here with Interchain Queries. A couple questions for my understanding.

1. Is it intended that the Publication by storing the KV pair in the IBC store or just by emitting an event whenever a particular key in the general state gets updated?
2. Is it the intention that the subscribing chain gets every single update to the value?
3. How would negotiation on data format work? Would it be implicit like current interchain queries.
4. Building on @ethanfrey 's comment, how does the subscribing chain respond if it receives update `[B, D]` but knows that there exists an update `C` in the middle which it has not yet received?

Unlike current Interchain queries, it should be noted that this will require changes to be implemented on the Publishing (Queried) chain since they must also store the height at which value updates as well as height of previous update.

Note that we can build a unidirectional PubSub on top of the bidirectional channel abstraction. We do something vey similar for ICA, where messages flow in one direction

**Reply by michaelfig** [reply]
> Thanks for the idea @michaelfig! Agreed with @zmanian that we should try to merge efforts here with Interchain Queries. A couple questions for my understanding.

BTW, I'm rebranding this as "streams".  Please see https://github.com/Agoric/agoric-sdk/blob/mfig-vstream/golang/cosmos/x/vstream/spec/01_concepts.md#concepts

> 1. Is it intended that the Publication by storing the KV pair in the IBC store or just by emitting an event whenever a particular key in the general state gets updated?

Storage is expected.

> 2. Is it the intention that the subscribing chain gets every single update to the value?

If it desires.  For some streams, it may be desirable to skip to the latest value when possible.

> 3. How would negotiation on data format work? Would it be implicit like current interchain queries.

Yes, it would be implicit.

> 4. Building on @ethanfrey 's comment, how does the subscribing chain respond if it receives update `[B, D]` but knows that there exists an update `C` in the middle which it has not yet received?

It issues an interchain query for that specific update.
 
> Unlike current Interchain queries, it should be noted that this will require changes to be implemented on the Publishing (Queried) chain since they must also store the height at which value updates as well as height of previous update.

Yes, it is intended as a layer on top of queries, not a replacement.
 
> Note that we can build a unidirectional PubSub on top of the bidirectional channel abstraction. We do something vey similar for ICA, where messages flow in one direction

That doesn't work for me, since I want there to be no data flow (handshake or acks) from the client to the Publishing chain.

**Comment by jackzampolin** [comment]
This is very cool, will see how we can merge this with any ibc query efforts

**Comment by michaelfig** [comment]
I've updated the concepts document linked from the description to feature use cases prominently.

**Comment by michaelfig** [comment]
Inviting @albertchon to the conversation!
